### PR TITLE
Fix #55: Add org-file-tags to inherited tags list

### LINKED ISSUE #55: #+FILETAGS are ignored by tags-inherited
I've tried several styles to specify filetags, unfortunately none of them are working. Example file:
```
#+FILETAGS: :Peter:Boss:Secret:
#+FILETAGS: :Boss:
#+FILETAGS: Secret

* Meeting with the French group      :work:
** Summary by Frank                  :boss:notes:
*** TODO Prepare slides for him      :action:
``` 
Inheritance from headline works OK:
```
(org-ql-search "~/testfile.org"
  '(tags "work")) => 3 items
```
And none of the filetags are matched:
```
(org-ql-search "~/testfile.org"
  '(tags "Peter")) => 0 items
```
```
(org-ql-search "~/testfile.org"
  '(tags "Boss")) => 0 items
```
```
(org-ql-search "~/testfile.org"
  '(tags "Secret")) => 0 items
```

### PR DESCRIPTION
**Pull request by mskorzhinskiy**
Tested with all examples from the bug description.

### CODE REVIEW

**Comment by alphapapa** [comment]
Thanks, but that won't work properly.  You'll need to study the code in that function carefully and understand exactly how it works.  The list of file tags will need to be appended to the list of inherited tags in the proper place.  I'm not sure yet exactly where in the code the appending should be done, because since the function is recursive (up the outline path), it would be best to avoid doing it for every recursive call, if possible.

**Comment by mskorzhinskiy** [comment]
I'm not sure why exactly it wouldn't work properly. Care to share any contr-example?

I should've attached my explanations along with the commit. Sorry about that. Anyway, my thinking here was the following.

Why it would work? Upmost item in the recursion will always have zero inherited tags, since it does not have any upper heading. So at the upmost item, instead of always returning org-ql-nil, we will now return org-file-tags or org-ql-nil.

So in case there are non-nil org-file-tags, any middle item in the hierarchy will have non-nil inherited-tags (from the upmost item) and the upmost item will take its inherited tags from file tags.

There are few problems though.

What if there are no file tags in the file? In case there are also no tags in the hierarchy,  org-file-tags will be evaluated at every step of recursion (to nil). I though it is not a big problem, it just getting a value. But it could be avoided by checking current heading. I was just thinking that it is a bit more costly operation.

Another problem is tags that are explicitly excluded from the inheritance (or disabled  tag inheritance). According to documentation and some very inaccurate stares at the org-mode code, it looks like all these matters are already handled while setting org-file-tags. So I didn't checked any of that.

**Comment by alphapapa** [comment]
> Another problem is tags that are explicitly excluded from the inheritance (or disabled tag inheritance). According to documentation and some very inaccurate stares at the org-mode code, it looks like all these matters are already handled while setting org-file-tags. So I didn't checked any of that.

I think this is the first problem to be solved.

1.  Even if Org leaves `org-file-tags` nil when `org-use-tag-inheritance` is nil, we shouldn't rely on that behavior going forward.  I can imagine the bug reports when that code in Org is refactored someday; I keep having to make this case on the Org mailing list when people propose changes without consideration for third-party packages that use the code.

2.  If `org-use-tags-inheritance` is bound to nil during an `org-ql` query, we must not use inherited tags, even if `org-file-tags` is non-nil.

3.  Also, we must consider `org-tags-exclude-from-inheritance`.  I guess there is no official algorithm for combining these options, but we should try to figure out what is best and do that.  In some cases that may mean copying what Org currently does, but that may not always be best.

It would be a very elegant patch if that one symbol in the `or` solved it, but I'm not sure it's that simple.  Org can be complicated.  :)

**Comment by sf-rasmi** [comment]
Yes, you're right, actually documentation of the org-file-tags variable explicitly says that 
>  List of tags that **can** be inherited by all entries in the file.
And actually running queries to select tagged items with disabled (or limited) tag inheritance and populated file tags, proves that these settings are not affecting org-file-tags variables.

That was wrong not to check it more thoroughly. But I just hoped that the patch would be really that elegant, yes. :-)

**Comment by mskorzhinskiy** [comment]
Yes, you're right, actually documentation of the org-file-tags variable explicitly says that
> List of tags that **can** be inherited by all entries in the file.

And actually running queries to select tagged items with disabled (or limited) tag inheritance and populated file tags, proves that these settings are not affecting org-file-tags variable.

That was wrong not to check it more thoroughly. But I just hoped that the patch would be really that elegant, yes. :-)

**Comment by alphapapa** [comment]
> Yes, you're right, actually documentation of the org-file-tags variable explicitly says that
> 
> > List of tags that **can** be inherited by all entries in the file.
> 
> And actually running queries to select tagged items with disabled (or limited) tag inheritance and populated file tags, proves that these settings are not affecting org-file-tags variable.

Thanks for testing that.  So now, unfortunately, we have to decide what the proper thing to do is:

1.  Imitate Org by including `org-file-tags` unconditionally.  This seems wrong to me, and may be considered a bug in Org, but fixing it would change established behavior.

2.  Apply the options in what seems like the logical way: only inherit file tags when tag inheritance is enabled.  In this case, we also have to decide whether to exclude file tags which are in `org-tags-exclude-from-inheritance`: in a way it seems nonsensical to exclude file tags from inheritance, but if a user wanted to do so for a particular search, that would seem like the way to do it, so it should probably be supported.

While I mostly want to imitate Org, in some ways `org-ql` is intended to be an improvement, and I think we should do what seems more correct rather than imitate Org here.  What do you think?

> That was wrong not to check it more thoroughly. But I just hoped that the patch would be really that elegant, yes. :-)

It could be fine, depending on one's intentions.  :)  I'd just rather go the extra mile here and do it "right."

Thanks for your work on this.

**Comment by mskorzhinskiy** [comment]
> While I mostly want to imitate Org, in some ways org-ql is intended to be an improvement, and I think we should do what seems more correct rather than imitate Org here. What do you think?

Surely inheritance rules should be treated inside org-ql. I was just too lazy to check how actually org-file-tags is populated with different settings. My bad.

So, I've tried to test with the same file (see in the bug description) but with different settings:
```
(let ((org-use-tag-inheritance (list "Boss")))
  (org-ql-search "~/file.org"
    '(tags "Peter"))) => nothing

(let ((org-use-tag-inheritance (list "Boss")))
  (org-ql-search "~/file.org"
    '(tags "Boss"))) => 3 headings

(let ((org-use-tag-inheritance (list "notes")))
  (org-ql-search "~/file.org"
    '(tags "notes"))) => 2 headings

(let ((org-use-tag-inheritance (list "abrakadabra")))
  (org-ql-search "~/file.org"
    '(tags "notes"))) => 1 heading

(let ((org-use-tag-inheritance nil))
  (org-ql-search "~/file.org"
    '(tags "notes"))) => noting
```

**Comment by alphapapa** [comment]
Thanks for your patience again.

We're modifying the `--tags-at` function here.  I plan to move to using `--value-at` for tags searches in the future, so I think we should make this change after doing that.  I want to tag 0.4 first, so I'm going to hold off on this PR until starting on 0.5-pre.

When that time comes, some refactoring will be required, so I doubt this PR will directly apply then.  I'll leave it open for now, and make a few more review comments for your reference, but be advised that it probably won't end up being merged as-is.  :)

However, whether or not the code ends up being merged as-is, you have helped very much with your analysis and testing, which I thank you for.  Please be sure that I credit you in the docs when the bug is fixed!

File: org-ql.el [hunk 1]
```diff
@@ -439,6 +438,17 @@ Returns cons (INHERITED-TAGS . LOCAL-TAGS)."
                   (list heading))
         (list heading)))))
 
+(defun org-ql--tags-apply-inheritence (tags)
```
Comment (alphapapa) [inline org-ql.el]: Typo: it's `inheritance`.  :)

File: org-ql.el [hunk 2]
```diff
@@ -439,6 +438,17 @@ Returns cons (INHERITED-TAGS . LOCAL-TAGS)."
...
+  "Apply inheritence settings on TAGS."
+  (cl-typecase org-use-tag-inheritance
+    (list (setf tags (-intersection tags org-use-tag-inheritance)))
+    (string (setf tags (--select (string-match org-use-tag-inheritance it)
+                                 tags))))
+  (pcase org-tags-exclude-from-inheritance
+    ('nil tags)
+    (_ (-difference tags org-tags-exclude-from-inheritance)))
+  tags)
```
Comment (alphapapa) [inline org-ql.el]: Note that the `pcase` should be the final form so that its value is returned.

**Comment by mskorzhinskiy** [comment]
I am not sure that I am the one whom you should be thanking. Thank you for *your* patience and reviewing me. Just to state the obvious — my major goal here is to have a bug fixed, so feel free to modify or totally rewrite my PR. I am not really care about authorship here.

However, I don't like the idea of postponing this bug til next refactoring. From my perspective this bug is rather easy to fix, test and very frustraiting for people with attached file tags (e.g. me).

Please consider applying something about #55 to 0.4.

**Comment by alphapapa** [comment]
Understood.  I don't use filetags much, so it didn't seem as important to me, but I'm not the only user.  :)  I'll fix it before the refactor.

**Comment by mskorzhinskiy** [comment]
Again — many thanks for your time and fixing this bug.

**Pull request closed without merging**